src/mscorlib/src/System/Guid.cs

```diff
@@ -365,214 +365,295 @@ public static bool TryParseExact(String input, String format, out Guid result)
             }
         }
 
-
-        private static bool TryParseGuid(String g, GuidStyles flags, ref GuidResult result)
+        private static unsafe bool TryParseGuid(string g, GuidStyles flags, ref GuidResult result)
         {
-            if (g == null) {
-                result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
-                return false;
-            }
-            String guidString = g.Trim();  //Remove Whitespace
-
-            if (guidString.Length == 0) {
+            if (g == null)
+            {
                 result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
                 return false;
             }
 
-            // Check for dashes
-            bool dashesExistInString = (guidString.IndexOf('-', 0) >= 0);
-
-            if (dashesExistInString) {
-                if ((flags & (GuidStyles.AllowDashes | GuidStyles.RequireDashes)) == 0) {
-                    // dashes are not allowed
-                    result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
-                    return false;
+            fixed (char* guidString = g)
+            {
+                int start = 0;
+                int end = g.Length;
+                for (int i = 0; i < end; i++)
+                {
+                    if (!char.IsWhiteSpace(guidString[i]))
+                    {
+                        break;
+                    }
+                    start++;
                 }
-            }
-            else {
-                if ((flags & GuidStyles.RequireDashes) != 0) {
-                    // dashes are required
+                for (int i = end; i > start; i--)
```

Comment:
This can be
`for (int i = end - 1; i > start; i--)`
instead, since `guidString[end]` is guaranteed to be '\0'.
However, I think it would be more elegant to these two loops as
```
int start = 0;
while (start < g.Length && char.IsWhitespace(guidString[start]))
  start++;

int end = g.Length;
while (end > start && char.IsWhitespace(guidString[end - 1]))
  end--;
```